`grafana_organization` data source returns member user data

mdb opened the pull request:
This seeks to address issue #480.

Specifically, its implementation is inspired by the following comment from @adkinsjd:

https://github.com/grafana/terraform-provider-grafana/issues/480#issuecomment-1127156583

The generated docs included in the PR -- as well as my PR annotation comments -- provide additional implementation notes and reviewer considerations.

Thanks!

mdb committed:
add `grafana_organization_users` data source

This seeks to address issue #480.

Specifically, it's implementation is inspired by the following comment
from @adkinsjd:

https://github.com/grafana/terraform-provider-grafana/issues/480#issuecomment-1127156583

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb commented on docs/data-sources/organization_users.md:
```diff
@@ -0,0 +1,43 @@
+
+data "grafana_organization_users" "from_name" {
+  organization_name = grafana_organization.test_org.name
+}
```
In theory, the already-existing `data.grafana_organization` could be enhanced to report user data, but I opted to make a separate, dedicated `grafana_organization_users` data source given that `GET /api/org/<id>` (used by `data.grafana_organization`) and `GET /api/org/<id>/users` are separate Grafana API endpoints. Do others agree?

julienduchesne replied:
I would make it a single resource. And mirror the attributes from the existing organization resource (admins, viewers, editors)

mdb replied:
> I would make it a single resource. And mirror the attributes from the existing organization resource (admins, viewers, editors)

@julienduchesne Thanks for the feedback! I'm amenable to that! But, before I get started, does it change your opinion to consider that...

1. That isn't how the [Grafana API itself](https://grafana.com/docs/grafana/latest/developers/http_api/org/#get-users-in-organization) models this data (`GET /api/org/<id>/users` is a separate endpoint & resource; the endpoint's associated JSON resource itself is simply a `[]OrgUser`)
2. At a glance, it seems other, existing `terraform-provider-grafana` data sources may err on the side of multiple, distinct, simple individual resources (for example, `data.grafana_team` returns no team membership/user attributes; instead, the `data.grafana_user` is used to fetch these attributes)

...or do you still favor a single `data.grafana_organization` data source with `admins`, `editors`, and `viewers` attributes?

julienduchesne replied:
Better to mirror the existing organization resource IMO

mdb replied:
@julienduchesne Sounds good and thanks for the feedback! Stay tuned...

mdb commented on grafana/data_source_organization_users.go:
```diff
@@ -0,0 +1,62 @@
+				Required:    true,
+				Description: "The name of the Organization.",
+			},
+			"emails": {
```
I considered calling this `user_emails` and/or `user_email_addresses`, though `emails` seems the most terse, clear, and consistent with the `email` attribute used elsewhere by the API and provider. Is that reasonable?

mdb commented on grafana/data_source_organization_users.go:
```diff
@@ -0,0 +1,62 @@
+					Type: schema.TypeString,
+				},
+				Description: "A list of the Organization member users' email addresses.",
+			},
```
I considered enhancing the data source with a `users` property:

```golang
"users": {
  Type:     schema.TypeSet,
  Computed: true,
  Elem: &schema.Resource{
    Schema: map[string]*schema.Schema{
      "email": {
        Type:     schema.TypeString,
        Computed: true,
      },
      "login": {
        Type:     schema.TypeString,
        Computed: true,
      },
      "role": {
        Type:     schema.TypeString,
        Computed: true,
      },
      "id": {
        Type:     schema.TypeInt,
        Computed: true,
      },
    },
  },
```

Would others like to see that as part of this PR now? If so, should that replace or coincide with the existing `emails`?

mdb commented on grafana/data_source_organization_users_test.go:
```diff
@@ -0,0 +1,61 @@
+	})
+}
+
+func testAccDatasourceOrganizationUsersConfig(orgName, email string) string {
```
I opted to use this configuration in tests -- rather than the `examples/data-sources/grafana_organization_users/data-source.tf` configuration -- so we can dynamically declare the `foo@example.com` user email address from within the tests, rather than needing to hard-code the expected email addresses across both the `examples/data-sources/grafana_organization_users/data-source.tf` and `grafana/data_source_organization_users_test.go` files.

Do others agree with this decision, or would you prefer I abandon `testAccDatasourceOrganizationUsersConfig` in favor of using a `examples/data-sources/grafana_organization_users/data-source.tf` configuration?

mdb committed:
organization data source has admins/editors/viewers

This adds the following attributes -- each a list of corresponding
user email addresses -- to the `data_source_organization`:`

* `admins`
* `editors`
* `viewers`

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
update data_source_organization docs

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
fine tune code formatting

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
test new organization data source attributes

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
correct data source name in tests

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
re-generate docs

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
remove data_source_organization_users

Instead, the data_source_organization now returns user data.

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
remove organization_users data source example

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb commented:
@julienduchesne I believe this is ready for re-review; thanks!

julienduchesne commented on grafana/data_source_organization.go:
```diff
@@ -38,6 +63,28 @@ func dataSourceOrganizationRead(ctx context.Context, d *schema.ResourceData, met
+
+	for roleCollection, emails := range userCollections {
+		if err := d.Set(roleCollection, emails); err != nil {
+			return diag.FromErr(fmt.Errorf("error setting %s: %v", roleCollection, err))
```
```suggestion
			return diag.Errorf("error setting %s: %v", roleCollection, err)
```

julienduchesne commented on grafana/data_source_organization.go:
```diff
@@ -38,6 +63,28 @@ func dataSourceOrganizationRead(ctx context.Context, d *schema.ResourceData, met
+		userCollections[role] = append(userCollections[role], user.Email)
+	}
+
+	for roleCollection, emails := range userCollections {
```
```suggestion
	for role, emails := range userCollections {
```

It's just a role, not a collection of it

mdb replied:
@julienduchesne I believe `roleCollection` refers to the key name of the `map[string][]string` and is plural (i.e. `admins`, `editors`, `viewers`). Nonethless, I'm happy to change this variable name if you prefer.

julienduchesne reviewed (approved):
LGTM! Just some nits

mdb committed:
Update grafana/data_source_organization.go

Co-authored-by: Julien Duchesne <julienduchesne@live.com>

mdb committed:
Update grafana/data_source_organization.go

Co-authored-by: Julien Duchesne <julienduchesne@live.com>

mdb committed:
fix compilation

Signed-off-by: Mike Ball <mikedball@gmail.com>

julienduchesne commented:
Thanks @mdb !
